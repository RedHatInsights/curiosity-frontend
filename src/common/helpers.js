/**
 * Generate a random'ish ID.
 *
 * @param {string} prefix
 * @returns {string}
 */
const generateId = prefix =>
  `${prefix || 'generatedid'}-${(process.env.REACT_APP_ENV !== 'test' && Math.ceil(1e5 * Math.random())) || ''}`;

// ToDo: expand to include "async" check in scenarios where async/await are utilized.
/**
 * Check if "is a Promise".
 *
 * @param {Promise|*} obj
 * @returns {boolean}
 */
const isPromise = obj => Object.prototype.toString.call(obj) === '[object Promise]';

/**
 * An empty function.
 * Typically used as a default prop.
 */
const noop = Function.prototype;

/**
 * An empty promise.
 * Typically used as a default prop, or during testing.
 *
 * @type {Promise<{}>}
 */
const noopPromise = Promise.resolve({});

/**
 * A placeholder for "t", translation method.
 * Associated with the i18n package, and typically used as a default prop.
 *
 * @param {string} key
 * @param {string|object} value
 * @param {Array} components
 * @returns {string}
 */
const noopTranslate = (key, value, components) =>
  `t(${key}${(value && `, ${value}`) || ''}${(components && `, ${components}`) || ''})`;

/**
 * Is dev mode active.
 * Associated with using the NPM script "start". See dotenv config files for activation.
 *
 * @type {boolean}
 */
const DEV_MODE = process.env.REACT_APP_ENV === 'development';

/**
 * Is prod mode active.
 * Associated with production builds. See dotenv config files for activation.
 *
 * @type {boolean}
 */
const PROD_MODE = process.env.REACT_APP_ENV === 'production';

/**
 * Is review/proxy mode active.
 * Associated with using the NPM script "start:proxy". See dotenv config files for activation.
 *
 * @type {boolean}
 */
const REVIEW_MODE = process.env.REACT_APP_ENV === 'review';

/**
 * Is test mode active.
 * Associated with running unit tests. See dotenv config files for activation.
 *
 * @type {boolean}
 */
const TEST_MODE = process.env.REACT_APP_ENV === 'test';

/**
 * Apply a path prefix for routing.
 * Typically associated with applying a "beta" path prefix. See dotenv config files for updating. See build scripts for generated prefix.
 *
 * @type {string}
 */
const UI_DEPLOY_PATH_PREFIX = process.env.REACT_APP_UI_DEPLOY_PATH_PREFIX;

/**
 * Disable an aspect of the UI.
 * Typically associated with disabling views through route settings. See dotenv config files for activation.
 *
 * @type {boolean}
 */
const UI_DISABLED = process.env.REACT_APP_UI_DISABLED === 'true';

/**
 * Disable the graph card aspect of the UI.
 * See dotenv config files for activation.
 *
 * @type {boolean}
 */
const UI_DISABLED_GRAPH = process.env.REACT_APP_UI_DISABLED_GRAPH === 'true';

/**
 * Disable the inventory/table aspect of the UI.
 * See dotenv config files for activation.
 *
 * @type {boolean}
 */
const UI_DISABLED_TABLE = process.env.REACT_APP_UI_DISABLED_TABLE === 'true';

/**
 * Disable the filter toolbar aspect of the UI.
 * See dotenv config files for activation.
 *
 * @type {boolean}
 */
const UI_DISABLED_TOOLBAR = process.env.REACT_APP_UI_DISABLED_TOOLBAR === 'true';

/**
 * UI application name.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_DISPLAY_NAME = process.env.REACT_APP_UI_DISPLAY_NAME;

/**
 * UI application configuration name.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_DISPLAY_CONFIG_NAME = process.env.REACT_APP_UI_DISPLAY_CONFIG_NAME;

/**
 * UI application sentence start name.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_DISPLAY_START_NAME = process.env.REACT_APP_UI_DISPLAY_START_NAME;

/**
 * UI locale default.
 *
 * @type {string}
 */
const UI_LOCALE_DEFAULT = process.env.REACT_APP_CONFIG_SERVICE_LOCALES_DEFAULT_LNG;

/**
 * UI locale default description.
 *
 * @type {string}
 */
const UI_LOCALE_DEFAULT_DESC = process.env.REACT_APP_CONFIG_SERVICE_LOCALES_DEFAULT_LNG_DESC;

/**
 * UI state logging name/id.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_LOGGER_ID = process.env.REACT_APP_UI_LOGGER_ID || 'GUI';

/**
 * UI packaged application name.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_NAME = process.env.REACT_APP_UI_NAME;

/**
 * UI packaged application path, with generated prefix.
 * See dotenv config files for updating. See build scripts for generated prefix.
 *
 * @type {string}
 */
const UI_PATH = process.env.PUBLIC_URL || '/';

/**
 * UI packaged application version, with generated hash.
 * See dotenv config files for updating. See build scripts for generated hash.
 *
 * @type {string}
 */
const UI_VERSION = process.env.REACT_APP_UI_VERSION;

/**
 * UI exposed window name/id.
 * See dotenv config files for updating.
 *
 * @type {string}
 */
const UI_WINDOW_ID = process.env.REACT_APP_UI_WINDOW_ID || 'GUI';

/**
 * Expose an application specific object.
 * Associated with access on a browser's developer console. Limits exposed additions to
 * test and non-production environments only. Exposes helpers across all environments.
 *
 * @param {object} obj
 * @param {object} options
 * @property {boolean} limit
 * @property {string} id
 */
const browserExpose = (obj = {}, options) => {
  const { limit = PROD_MODE, id = UI_WINDOW_ID } = options || {};
  window[id] = (limit && { ...window[id] }) || { ...window[id], ...obj };
};

const helpers = {
  browserExpose,
  generateId,
  isPromise,
  noop,
  noopPromise,
  noopTranslate,
  DEV_MODE,
  PROD_MODE,
  REVIEW_MODE,
  TEST_MODE,
  UI_DEPLOY_PATH_PREFIX,
  UI_DISABLED,
  UI_DISABLED_GRAPH,
  UI_DISABLED_TABLE,
  UI_DISABLED_TOOLBAR,
  UI_DISPLAY_NAME,
  UI_DISPLAY_CONFIG_NAME,
  UI_DISPLAY_START_NAME,
  UI_LOCALE_DEFAULT,
  UI_LOCALE_DEFAULT_DESC,
  UI_LOGGER_ID,
  UI_NAME,
  UI_PATH,
  UI_VERSION,
  UI_WINDOW_ID
};

/**
 * Expose helpers to the browser's developer console.
 */
helpers.browserExpose({ ...helpers }, { limit: false });

export { helpers as default, helpers };
