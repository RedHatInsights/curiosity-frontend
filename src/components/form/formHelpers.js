/**
 * Confirm a string has minimum length.
 *
 * @param {string} value
 * @param {number} characters
 * @returns {boolean}
 */
const doesNotHaveMinimumCharacters = (value, characters = 1) =>
  (typeof value === 'string' && value.length < characters) || typeof value !== 'string';

const formHelpers = {
  doesNotHaveMinimumCharacters
};

export { formHelpers as default, formHelpers, doesNotHaveMinimumCharacters };
