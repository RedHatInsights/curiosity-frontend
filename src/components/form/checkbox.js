import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox as PfCheckbox } from '@patternfly/react-core/dist/js/components/Checkbox';
import { helpers } from '../../common';

/**
 * Render a checkbox form element.
 *
 * @param {object} props
 * @param {string} props.ariaLabel
 * @param {*} props.checked
 * @param {Node} props.children
 * @param {boolean} props.isDisabled
 * @param {Node} props.label
 * @param {string} props.name
 * @param {Function} props.onChange
 * @param {boolean} props.readOnly
 * @param {*} props.value
 * @returns {Node}
 */
const Checkbox = ({ ariaLabel, checked, children, isDisabled, label, name, onChange, readOnly, value, ...props }) => {
  const nameId = name || helpers.generateId();

  const onCheckboxChange = (isChecked, event) => {
    const { currentTarget, target } = event;
    const mockEvent = {
      id: nameId,
      name: nameId,
      value,
      checked: isChecked,
      target,
      currentTarget,
      persist: helpers.noop
    };
    onChange(mockEvent);
  };

  return (
    <PfCheckbox
      aria-label={ariaLabel || children || label}
      checked={checked}
      id={nameId}
      isChecked={checked}
      isDisabled={isDisabled || readOnly || false}
      label={children || label}
      name={nameId}
      onChange={onCheckboxChange}
      value={value}
      readOnly={readOnly || false}
      {...props}
    />
  );
};

/**
 * Prop types.
 *
 * @type {{onChange: Function, children: Node, name: string, checked: boolean, readOnly: boolean,
 *     label: string, value: *, ariaLabel: string}}
 */
Checkbox.propTypes = {
  ariaLabel: PropTypes.string,
  checked: PropTypes.any,
  children: PropTypes.node,
  isDisabled: PropTypes.bool,
  label: PropTypes.node,
  name: PropTypes.string,
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
  value: PropTypes.any
};

/**
 * Default props.
 *
 * @type {{onChange: Function, children: null, name: null, checked: undefined, readOnly: undefined,
 *     label: string, value: undefined, ariaLabel: null}}
 */
Checkbox.defaultProps = {
  ariaLabel: null,
  checked: undefined,
  children: null,
  isDisabled: false,
  label: '',
  name: null,
  onChange: helpers.noop,
  readOnly: false,
  value: undefined
};

export { Checkbox as default, Checkbox };
