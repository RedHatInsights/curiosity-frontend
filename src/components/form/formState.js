import React from 'react';
import PropTypes from 'prop-types';
import _cloneDeep from 'lodash/cloneDeep';
import _isEqual from 'lodash/isEqual';
import helpers from '../../common/helpers';

/**
 * Initial component state values.
 *
 * @type {{submitCount: number, isValid: null, isUpdating: boolean, isSubmitting: boolean, isValidating: boolean}}
 */
const initialState = {
  isUpdating: false,
  isSubmitting: false,
  isValid: null,
  isValidating: false,
  submitCount: 0
};

/**
 * Maintain a form's state/context.
 *
 * @augments React.Component
 * @fires onEvent
 * @fires onEventCustom
 * @fires onReset
 * @fires onSubmit
 */
class FormState extends React.Component {
  /**
   * Infer a field value is "checked" from a boolean value.
   *
   * @param {object} params
   * @param {boolean} params.setValuesAssumeBoolIsChecked
   * @param {object} params.setValues
   * @returns {{}}
   */
  static checkedSetValues({ setValuesAssumeBoolIsChecked, setValues }) {
    const checked = {};

    if (setValuesAssumeBoolIsChecked) {
      Object.keys(setValues).forEach(key => {
        if (typeof setValues[key] === 'boolean') {
          checked[key] = setValues[key];
        }
      });
    }

    return checked;
  }

  /**
   * Check if "is a promise".
   *
   * @param {*} obj
   * @returns {boolean}
   */
  static isPromise(obj) {
    return Object.prototype.toString.call(obj) === '[object Promise]';
  }

  state = {
    ...initialState
  };

  constructor(props) {
    super(props);

    this.touched = {};
    this.checked = FormState.checkedSetValues(props);

    this.refValues =
      props.resetUsingSetValues === true || props.setValuesOnUpdate === true ? _cloneDeep(props.setValues) : null;

    this.errors = {};
    this.values = _cloneDeep(props.setValues);
  }

  componentDidMount() {
    const { validateOnMount } = this.props;

    if (validateOnMount === true) {
      this.validateOnMount();
    }
  }

  componentDidUpdate() {
    const { refValues } = this;
    const { setValuesOnUpdate, setValues } = this.props;

    if (setValuesOnUpdate === true && !_isEqual(refValues, setValues)) {
      this.updateComponentValues();
    }
  }

  /**
   * Apply form values with a custom event.
   *
   * @event onEventCustom
   * @param {Array|object} custom
   */
  onEventCustom = custom => {
    const eventArray = (Array.isArray(custom) && custom) || (custom && [custom]);

    if (!eventArray.length) {
      return;
    }

    eventArray
      .filter(event => 'name' in event && ('value' in event || 'checked' in event))
      .forEach(event => this.onEvent({ target: { ...event }, persist: helpers.noop, type: 'custom' }));
  };

  /**
   * Generic "on event" for handling returned event objects.
   *
   * @event onEvent
   * @param {object} event
   */
  onEvent = event => {
    const { touched, values } = this;
    const { id, name, value, checked } = event.options ? { ...event } : event.target;

    event.persist();

    const targetName = name || id || 'generated form state target, add name or id attr to field';

    this.touched = { ...touched, [targetName]: true };
    this.values = { ...values, [targetName]: value };

    if (checked !== undefined) {
      this.checked = { ...this.checked, [targetName]: checked };
    }

    this.setState(
      {
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.values(setUpdateErrors).filter(val => val === true).length;

          this.errors = setUpdateErrors;

          this.setState({
            isUpdating: false,
            isValid: checkIsValid,
            isValidating: false
          });
        })
    );
  };

  /**
   * Reset FormState's form state. Apply prop onReset function.
   *
   * @event onReset
   * @param {object} event
   */
  onReset = event => {
    const { refValues, values } = this;
    const { setValuesAssumeBoolIsChecked, onReset, resetUsingSetValues } = this.props;

    event.persist();

    const isResetWithSetValues = refValues && resetUsingSetValues === true;
    const updatedValues = (isResetWithSetValues && _cloneDeep(refValues)) || {};
    const updatedChecked =
      (isResetWithSetValues && FormState.checkedSetValues(setValuesAssumeBoolIsChecked, updatedValues)) || {};

    this.values = updatedValues;
    this.checked = updatedChecked;
    this.errors = {};
    this.touched = {};

    this.setState({
      ...initialState
    });

    if (isResetWithSetValues) {
      onReset({ event, ..._cloneDeep({ values: updatedValues, prevValues: values }) });
    } else {
      // Resetting the values, potentially, will throw the controlled vs uncontrolled messaging.
      onReset({ event, values: {}, ..._cloneDeep({ prevValues: values }) });
    }
  };

  /**
   * Validate form, then submit.
   *
   * @event onSubmit
   * @param {object} event
   */
  onSubmit = event => {
    const { submitCount } = this.state;

    event.persist();
    event.preventDefault();

    this.setState(
      {
        submitCount: submitCount + 1,
        isSubmitting: true,
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.values(setUpdateErrors).filter(val => val === true).length;

          this.errors = setUpdateErrors;
          this.touched = {};

          this.setState(
            {
              isValid: checkIsValid,
              isValidating: false
            },
            () =>
              checkIsValid &&
              this.submit(event).then(() => {
                this.setState({
                  isSubmitting: false,
                  isUpdating: false
                });
              })
          );
        })
    );
  };

  /**
   * Handle submitted form, check and return Promise, or emulate for consistency.
   *
   * @param {object} event
   * @returns {Promise}
   */
  submit(event = { type: 'submit' }) {
    const { checked, errors, values, touched } = this;
    const { onSubmit } = this.props;

    const checkPromise = onSubmit({
      event,
      ..._cloneDeep({ ...this.state, checked, errors, values, touched })
    });

    if (FormState.isPromise(checkPromise)) {
      return checkPromise;
    }

    return {
      then: callback => callback()
    };
  }

  /**
   * Handle validated form data, check and return Promise, or emulate for consistency.
   *
   * @param {object} event
   * @returns {Promise}
   */
  validate(event = { type: 'validate' }) {
    const { checked, errors, values, touched } = this;
    const { validate } = this.props;

    const checkPromise = validate({
      event,
      ..._cloneDeep({ ...this.state, checked, errors, values, touched })
    });

    if (FormState.isPromise(checkPromise)) {
      return checkPromise;
    }

    return {
      then: callback => callback(checkPromise)
    };
  }

  /**
   * Shortcut, activate validation on component mount.
   *
   * @param {object} event
   */
  validateOnMount(event = { type: 'mount' }) {
    this.validateOnUpdate(event);
  }

  /**
   * Validate on component update.
   *
   * @param {object} event
   */
  validateOnUpdate(event = { type: 'update' }) {
    this.setState(
      {
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setUpdateErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.values(setUpdateErrors).filter(val => val === true).length;

          this.errors = setUpdateErrors;

          this.setState({
            isUpdating: false,
            isValidating: false,
            isValid: checkIsValid
          });
        })
    );
  }

  /**
   * On component update, update state.
   */
  updateComponentValues() {
    const { setValues, setValuesAssumeBoolIsChecked, validateOnUpdate } = this.props;

    this.setState(
      {
        isUpdating: true
      },
      () => {
        this.checked = FormState.checkedSetValues(setValuesAssumeBoolIsChecked, setValues);
        this.refValues = _cloneDeep(setValues);
        this.values = _cloneDeep(setValues);

        if (validateOnUpdate) {
          this.validateOnUpdate();
        } else {
          this.setState({
            isUpdating: false
          });
        }
      }
    );
  }

  /**
   * Pass child components, integrate and apply form context.
   *
   * @returns {Node}
   */
  render() {
    const { checked, errors, values, touched } = this;
    const { children } = this.props;

    return (
      <React.Fragment>
        {children({
          handleOnEventCustom: this.onEventCustom,
          handleOnEvent: this.onEvent,
          handleOnReset: this.onReset,
          handleOnSubmit: this.onSubmit,
          ..._cloneDeep({ ...this.state, checked, errors, values, touched })
        })}
      </React.Fragment>
    );
  }
}

/**
 * Prop types.
 *
 * @type {{setValuesOnUpdate: boolean, validateOnMount: boolean, children: Function,
 *     setValues: object, onSubmit: Function, resetUsingSetValues: boolean,
 *     setValuesAssumeBoolIsChecked: boolean, onReset: Function, validate: Function,
 *     validateOnUpdate: boolean}}
 */
FormState.propTypes = {
  children: PropTypes.func.isRequired,
  onReset: PropTypes.func,
  onSubmit: PropTypes.func,
  resetUsingSetValues: PropTypes.bool,
  setValues: PropTypes.object,
  setValuesOnUpdate: PropTypes.bool,
  setValuesAssumeBoolIsChecked: PropTypes.bool,
  validate: PropTypes.func,
  validateOnMount: PropTypes.bool,
  validateOnUpdate: PropTypes.bool
};

/**
 * Default props.
 *
 * @type {{setValuesOnUpdate: boolean, validateOnMount: boolean, setValues: {},
 *     onSubmit: Function, resetUsingSetValues: boolean, setValuesAssumeBoolIsChecked: boolean,
 *     onReset: Function, validate: Function, validateOnUpdate: boolean}}
 */
FormState.defaultProps = {
  onReset: helpers.noop,
  onSubmit: helpers.noop,
  resetUsingSetValues: true,
  setValues: {},
  setValuesOnUpdate: false,
  setValuesAssumeBoolIsChecked: true,
  validate: helpers.noop,
  validateOnMount: false,
  validateOnUpdate: false
};

export { FormState as default, FormState };
