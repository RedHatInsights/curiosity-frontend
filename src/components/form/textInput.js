import React from 'react';
import PropTypes from 'prop-types';
import { TextInput as PfTextInput } from '@patternfly/react-core';
import { createMockEvent } from './formHelpers';
import { helpers } from '../../common';

/**
 * Text input with state.
 *
 * @memberof Form
 * @module TextInput
 */

/**
 * A wrapper for Patternfly TextInput. Provides restructured event data,
 * and an onClear event for the search type.
 *
 * @augments React.Component
 * @fires onKeyUp
 * @fires onMouseUp
 * @fires onChange
 */
class TextInput extends React.Component {
  state = {
    updatedValue: null
  };

  /**
   * onKeyUp event, provide additional functionality for onClear event.
   *
   * @event onKeyUp
   * @param {object} event
   */
  onKeyUp = event => {
    const { onClear, onKeyUp, type } = this.props;
    const { currentTarget, keyCode } = event;
    const clonedEvent = { ...event };

    onKeyUp(createMockEvent(event, true));

    if (keyCode === 27) {
      if (type === 'search' && currentTarget.value === '') {
        onClear(createMockEvent(clonedEvent));
      } else {
        this.setState({ updatedValue: '' }, () => {
          onClear(
            createMockEvent({ ...clonedEvent, ...{ currentTarget: { ...clonedEvent.currentTarget, value: '' } } })
          );
        });
      }
    }
  };

  /**
   * onMouseUp event, provide additional functionality for onClear event.
   *
   * @event onMouseUp
   * @param {object} event
   */
  onMouseUp = event => {
    const { onClear, onMouseUp, type } = this.props;
    const { currentTarget } = event;
    const clonedEvent = { ...event };

    onMouseUp(createMockEvent(event, true));

    if (type !== 'search' || currentTarget.value === '') {
      return;
    }

    setTimeout(() => {
      if (currentTarget.value === '') {
        onClear(createMockEvent(clonedEvent));
      }
    });
  };

  /**
   * onChange event, provide restructured event.
   *
   * @event onChange
   * @param {string} value
   * @param {object} event
   */
  onChange = (value, event) => {
    const { onChange } = this.props;
    const clonedEvent = { ...event };

    this.setState({ updatedValue: value }, () => {
      onChange(createMockEvent(clonedEvent));
    });
  };

  /**
   * Render a text input.
   *
   * @returns {React.ReactNode}
   */
  render() {
    const { updatedValue } = this.state;
    const {
      className,
      id,
      isDisabled,
      name,
      onChange,
      onClear,
      onKeyUp,
      onMouseUp,
      isReadOnly,
      type,
      value,
      ...props
    } = this.props;
    const updatedName = name || helpers.generateId();
    const updatedId = id || updatedName;

    return (
      <PfTextInput
        id={updatedId}
        name={updatedName}
        className={`curiosity-text-input ${className}`}
        isDisabled={isDisabled || false}
        onChange={this.onChange}
        onKeyUp={this.onKeyUp}
        onMouseUp={this.onMouseUp}
        isReadOnly={isReadOnly || false}
        type={type}
        value={updatedValue ?? value ?? ''}
        {...props}
      />
    );
  }
}

/**
 * Prop types.
 *
 * @type {{onKeyUp: Function, isReadOnly: boolean, onChange: Function, onClear: Function,
 *     name: string, className: string, id: string, isDisabled: boolean, onMouseUp: Function,
 *     type: string, value: string}}
 */
TextInput.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  isDisabled: PropTypes.bool,
  isReadOnly: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  onClear: PropTypes.func,
  onKeyUp: PropTypes.func,
  onMouseUp: PropTypes.func,
  type: PropTypes.string,
  value: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{onKeyUp: Function, isReadOnly: boolean, onChange: Function, onClear: Function,
 *     name: string, className: string, id: string, isDisabled: boolean, onMouseUp: Function,
 *     type: string, value: string}}
 */
TextInput.defaultProps = {
  className: '',
  id: null,
  isDisabled: false,
  isReadOnly: false,
  name: null,
  onChange: helpers.noop,
  onClear: helpers.noop,
  onKeyUp: helpers.noop,
  onMouseUp: helpers.noop,
  type: 'text',
  value: ''
};

export { TextInput as default, TextInput };
