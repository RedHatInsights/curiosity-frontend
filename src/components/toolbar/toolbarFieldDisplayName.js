import React from 'react';
import PropTypes from 'prop-types';
import { InputGroup } from '@patternfly/react-core';
import _debounce from 'lodash/debounce';
import { reduxTypes, store, useSelector } from '../../redux';
import { TextInput } from '../form/textInput';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { translate } from '../i18n/i18n';

/**
 * ToDo: evaluate the debounce milliseconds, currently based off platforms default 800 ms
 */
/**
 * Display a display name input field for search.
 *
 * @fires onSubmit
 * @fires onClear
 * @fires onKeyUp
 * @param {object} props
 * @param {string} props.value
 * @param {Function} props.t
 * @param {string} props.viewId
 * @returns {Node}
 */
const ToolbarFieldDisplayName = ({ value, t, viewId }) => {
  const currentValue = useSelector(
    ({ view }) => view.inventoryHostsQuery?.[viewId]?.[RHSM_API_QUERY_TYPES.DISPLAY_NAME],
    value
  );

  /**
   * On submit, dispatch type.
   *
   * @event onSubmit
   * @param {string} submitValue
   * @returns {void}
   */
  const onSubmit = submitValue =>
    store.dispatch([
      {
        type: reduxTypes.query.SET_QUERY_CLEAR_INVENTORY_LIST
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_TYPES.DISPLAY_NAME],
        viewId,
        [RHSM_API_QUERY_TYPES.DISPLAY_NAME]: submitValue?.trim() || null
      }
    ]);

  /**
   * On clear, dispatch type.
   *
   * @event onClear
   * @returns {void}
   */
  const onClear = () => {
    if (currentValue === '' || !currentValue) {
      return;
    }

    store.dispatch([
      {
        type: reduxTypes.query.SET_QUERY_CLEAR_INVENTORY_LIST
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_TYPES.DISPLAY_NAME],
        viewId,
        [RHSM_API_QUERY_TYPES.DISPLAY_NAME]: null
      }
    ]);
  };

  /**
   * Set up submit debounce event to allow for bypass.
   */
  const debounced = _debounce(onSubmit, 700);

  /**
   * On enter submit value, on type submit value, and on esc ignore (clear value at component level).
   *
   * @event onKeyUp
   * @param {object} event
   */
  const onKeyUp = event => {
    switch (event.keyCode) {
      case 13:
        onSubmit(event.value);
        break;
      case 27:
        break;
      default:
        debounced(event.value);
        break;
    }
  };

  return (
    <InputGroup>
      <TextInput
        aria-label={t('curiosity-toolbar.placeholder', { context: 'displayName' })}
        className="curiosity-input__display-name"
        iconVariant="search"
        maxLength={255}
        onClear={onClear}
        onKeyUp={onKeyUp}
        value={currentValue}
        placeholder={t('curiosity-toolbar.placeholder', { context: 'displayName' })}
      />
    </InputGroup>
  );
};

/**
 * Prop types.
 *
 * @type {{viewId: string, t: Function, value: string}}
 */
ToolbarFieldDisplayName.propTypes = {
  t: PropTypes.func,
  value: PropTypes.string,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{viewId: string, t: translate, value: string}}
 */
ToolbarFieldDisplayName.defaultProps = {
  t: translate,
  value: null,
  viewId: 'toolbarFieldDisplayName'
};

export { ToolbarFieldDisplayName as default, ToolbarFieldDisplayName };
