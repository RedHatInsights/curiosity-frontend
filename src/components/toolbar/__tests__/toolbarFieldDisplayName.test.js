import React from 'react';
import { mount, shallow } from 'enzyme';
import { ToolbarFieldDisplayName } from '../toolbarFieldDisplayName';
import { store } from '../../../redux/store';
import { TextInput } from '../../form/textInput';
import { helpers } from '../../../common/helpers';

describe('ToolbarFieldDisplayName Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a non-connected component', () => {
    const props = {};
    const component = shallow(<ToolbarFieldDisplayName {...props} />);

    expect(component).toMatchSnapshot('non-connected');
  });

  it('should handle updating display name through redux state', () => {
    const props = {
      value: 'lorem ipsum'
    };

    const component = shallow(<ToolbarFieldDisplayName {...props} />);
    component.find(TextInput).simulate('change', { value: 'dolor sit' });
    component.find(TextInput).simulate('keyUp', {});

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch display name');
  });

  it('should handle updating display name using the enter key', () => {
    const props = {
      value: 'lorem ipsum'
    };

    const component = mount(<ToolbarFieldDisplayName {...props} />);
    const mockEvent = { keyCode: 13, currentTarget: { value: 'dolor sit' }, persist: helpers.noop };
    component.find(TextInput).instance().onKeyUp(mockEvent);

    expect(mockDispatch.mock.calls).toMatchSnapshot('enter display name');
  });

  it('should handle clearing display name through redux state', () => {
    const props = {
      value: 'lorem ipsum'
    };

    const component = mount(<ToolbarFieldDisplayName {...props} />);
    const mockEvent = { keyCode: 27, currentTarget: { value: '' }, persist: helpers.noop };
    component.find(TextInput).instance().onKeyUp(mockEvent);

    expect(mockDispatch.mock.calls).toMatchSnapshot('clear display name');
  });
});
