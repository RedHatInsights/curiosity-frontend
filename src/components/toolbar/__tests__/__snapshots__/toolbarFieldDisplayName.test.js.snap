// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarFieldDisplayName Component should handle clearing display name through redux state: clear display name 1`] = `
Array [
  Array [
    Array [
      Object {
        "type": "SET_QUERY_CLEAR_INVENTORY_LIST",
      },
      Object {
        "display_name_contains": null,
        "type": "SET_QUERY_RHSM_HOSTS_INVENTORY_display_name_contains",
        "viewId": "toolbarFieldDisplayName",
      },
    ],
  ],
]
`;

exports[`ToolbarFieldDisplayName Component should handle updating display name through redux state: dispatch display name 1`] = `
Array [
  Array [
    Array [
      Object {
        "type": "SET_QUERY_CLEAR_INVENTORY_LIST",
      },
      Object {
        "display_name_contains": "dolor sit",
        "type": "SET_QUERY_RHSM_HOSTS_INVENTORY_display_name_contains",
        "viewId": "toolbarFieldDisplayName",
      },
    ],
  ],
]
`;

exports[`ToolbarFieldDisplayName Component should handle updating display name using the enter key: enter display name 1`] = `
Array [
  Array [
    Array [
      Object {
        "type": "SET_QUERY_CLEAR_INVENTORY_LIST",
      },
      Object {
        "display_name_contains": "dolor sit",
        "type": "SET_QUERY_RHSM_HOSTS_INVENTORY_display_name_contains",
        "viewId": "toolbarFieldDisplayName",
      },
    ],
  ],
]
`;

exports[`ToolbarFieldDisplayName Component should render a non-connected component: non-connected 1`] = `
<InputGroup>
  <TextInput
    aria-label="t(curiosity-toolbar.placeholder, {\\"context\\":\\"displayName\\"})"
    className=""
    isDisabled={false}
    isReadOnly={false}
    name={null}
    onChange={[Function]}
    onClear={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    placeholder="t(curiosity-toolbar.placeholder, {\\"context\\":\\"displayName\\"})"
    type="search"
    value={null}
  />
  <Button
    aria-label="t(curiosity-toolbar.button, {\\"context\\":\\"displayName\\"})"
    onClick={[Function]}
    variant="control"
  >
    <SearchIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Button>
</InputGroup>
`;
